waiter/src/waiter/auth/authentication.clj

```diff
@@ -201,3 +219,62 @@
               handler' (middleware/wrap-merge handler auth-params-map)]
           (handler' request))
         (handler request)))))
+
+(defn process-auth-expires-at-request
+  "Handler to allow a client to update it's knowledge of when user's cookie-based credentials expire.
+   Returns a json response containing the expires-at key containing the expiration time in UTC epoch seconds.
+   Relies on the metadata in the x-waiter-auth cookie."
+  [password {:keys [headers]}]
+  (let [decoded-auth-cookie (get-and-decode-auth-cookie-value headers password)
+        [auth-principal _ auth-metadata] decoded-auth-cookie
+        {:keys [expires-at]} auth-metadata
+        expires-at (or (when (decoded-auth-valid? decoded-auth-cookie) expires-at) 0)]
+    (log/info auth-principal "cookie expires at" expires-at)
+    (utils/attach-waiter-source (utils/clj->json-response {:expires-at expires-at
+                                                           :principal auth-principal}))))
+
+(defn process-auth-keep-alive-request
+  "Handler to eagerly trigger authentication workflow even if cookie has not yet expired.
+   This allows clients to pre-emptively refresh credentials before they expire.
+   Presence of done parameter is used to avoid infinite auth redirect loops and will return 204 No Content.
+   Invalid offset parameters result in a 400 error.
+   Missing offset or soon to expire cookie (based on offset) will trigger in the authentication workflow.
+   If cookie is expected to be live longer than offset value, return 204 No Content."
+  [password auth-handler {:keys [headers] :as request}]
+  (let [{:strs [done offset]} (-> request ru/query-params-request :query-params)
```

Comment:
Can we call this `delta` instead of `offset`? I was confused about what this meant because I usually see "offset" used for string offsets, or for UTC offsets (i.e., timezones).

Reply:
😢